tools/awslint/GUIDELINES.md

```diff
@@ -0,0 +1,295 @@
+APIs in the AWS Construct Library in order to ensure a consistent and integrated
+experience across the entire AWS surface area.
+
+As much as possible, the guidelines in this document are enforced using the
+__awslint__ tool which reflects on the APIs and verifies that the APIs adhere to
+the guidelines.
+
+When a guideline is backed by a linter rule, the rule name will be referenced
+like this: <a id="resource-class-is-construct">`awslint|resource-class-is-construct`</a>
+and anchored with the rule name.
+
+For the purpose of this document we will use "**Xxx**" to denote the official
+name of the resource as defined in the AWS CloudFormation resource specification
+(i.e. "Bucket", "Queue", "Topic", etc). This notation allows deriving names from
+the official name. For example, `XxxProps` would be `BucketProps`, `TopicProps`,
+etc, `IXxx` would be `IBucket`, `ITopic` and so forth.
+
+The guidelines in this document use TypeScript (and npm package names) since
+this is the source programming language used to author the library, which is
+later packaged and published to all programming languages through
+[jsii](https://github.com/awslabs/jsii).
+
+## Modules
+
+> <a id="module-name">`awslint|module-name`</a>
+
+AWS resources are organized into modules based on their AWS service. For
+example, the "Bucket" resource, which is offered by the Amazon S3 service will
+be available under the **@aws-cdk/aws-s3** module.
+
+## Constructs
+
+Constructs are the basic building block of CDK applications. They represent
+abstract cloud resources of any complexity.
+
+> <a id="resource-class">awslint: construct-ctor</a>
+
+Construct initializer (constructor) signature should always be:
+
+```ts
+constructor(scope: cdk.Construct, id: string, props: XxxProps)
+```
+
+> <a id="resource-class">*awslint: construct-ctor-optional-props*</a>
+
+If all initialization properties are optional, the `props` argument must also be
+optional.
+
+```ts
+constructor(scope: cdk.Construct, id: string, props: XxxProps = { })
+```
+
+## CloudFormation Resource Classes
+
+Each module in the AWS Construct Library includes generated constructs which
+represent the "raw" CloudFormation resources.
+
+1. These classes are named `CfnXxx`
+2. They extend `cdk.Resource` (which extends `cdk.Construct`)
+3. Their constructor accepts a `props` parameter of type `CfnXxxProps`
+4. `CfnXxxProps` represents the exact set of properties as defined in the
+   AWS CloudFormation resource specification.
+5. They have readonly properties which represent all the runtime attributes of
+   the resource.
+
+> NOTE: there are no linting rules against this section because this layer
+> is entirely generated by the __cfn2ts__ tool according to these guidelines
+> so there is no need to lint against it.
+
+## AWS Resources
+
+> <a id="resource-class">`awslint|resource-class`</a>
+
+Each `CfnXxx` resource class must have a corresponding `Xxx` high-level (L2)
```

Comment:
I suppose this is to get some measure of completeness? I don't agree with this as a blanket rule; sometimes you want to hide a resource behind something else.

Reply:
Currently this rule is marked as a warning and we can make the call which L2s we implement and which we don't.